### PR TITLE
nixos/netdata: add configDir option

### PR DESCRIPTION
**Pull request by misuzu**
###### Motivation for this change
Resurrect https://github.com/NixOS/nixpkgs/pull/50048 with slightly modified logic (values in `configDir` are paths instead of lines).

###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by Izorkin** [comment]
If `netdata` service will write to this folder, then need to add parametr `serviceConfig.ReadWritePaths`.

**Comment by misuzu** [comment]
> If `netdata` service will write to this folder, then need to add parametr `serviceConfig.ReadWritePaths`.

Iit's readonly configuration directory, the only difference is we build our own instead of using `${pkgs.netdata}/lib/netdata/conf.d`.

File: nixos/modules/services/monitoring/netdata.nix [hunk 1]
```diff
@@ -130,6 +141,26 @@ in {
...
+          Each top-level attribute denotes a path in the configuration
+          directory as in environment.etc.
+          Its value is the absolute path and must be readable by netdata.
+          Cannot be combined with configText.
+        '';
+        example = literalExpression ''
+          "health_alarm_notify.conf" = pkgs.writeText "health_alarm_notify.conf" '''
+            sendmail="/path/to/sendmail"
+          ''';
+          "health.d" = "/run/secrets/netdata/health.d";
```
Comment (Izorkin) [inline nixos/modules/services/monitoring/netdata.nix:160]: Example will not work correctly. The folder `health.d` is created inside a subfolder `health.d`
Result:
```
l /nix/store/w3kv04ycchzz3r03ws1daf8knpfnca0n-netdata-config-d | grep health
lrwxrwxrwx    1 root root     68 янв  1  1970 health_alarm_notify.conf -> /nix/store/54m6scjd3hf1hqpcdym6fww8k6126g96-health_alarm_notify.conf
dr-xr-xr-x    2 root root   4,0K янв  1  1970 health.d
-r--r--r--    1 root root     57 янв  1  1970 health_email_recipients.conf

l /nix/store/w3kv04ycchzz3r03ws1daf8knpfnca0n-netdata-config-d/health.d | grep health
lrwxrwxrwx 1 root root   29 янв  1  1970 health.d -> /run/secrets/netdata/health.d
```
The intended result:
```
l /nix/store/5162yws7lnl3f9nsw5h9ixgafhr764zc-netdata-config-d | grep health
lrwxrwxrwx    1 root root     68 янв  1  1970 health_alarm_notify.conf -> /nix/store/54m6scjd3hf1hqpcdym6fww8k6126g96-health_alarm_notify.conf
lrwxrwxrwx    1 root root     29 янв  1  1970 health.d -> /run/secrets/netdata/health.d
-r--r--r--    1 root root     57 янв  1  1970 health_email_recipients.conf

```

**Comment by Izorkin** [comment]
Is it possible to add a symlink to a folder `/etc/netdata` ?

File: nixos/modules/services/monitoring/netdata.nix (hunk 1 above)
Comment (misuzu) [reply nixos/modules/services/monitoring/netdata.nix:160]: Fixed, thanks!

**Comment by misuzu** [comment]
> Is it possible to add a symlink to a folder `/etc/netdata` ?

Done.

File: nixos/modules/services/monitoring/netdata.nix [hunk 2]
```diff
@@ -19,8 +19,20 @@ let
     "${wrappedPlugins}/libexec/netdata/plugins.d"
   ] ++ cfg.extraPluginPaths;
 
+  configDirectory = pkgs.runCommand "netdata-config-d" { } ''
+    mkdir -p $out
+    cp -r ${cfg.package}/lib/netdata/conf.d/* $out
```
Comment (Izorkin) [inline nixos/modules/services/monitoring/netdata.nix:24]: No need to copy all config files:
```
2021-11-30 14:55:53: apps.plugin ERROR : MAIN : PROCFILE: Cannot open file '/etc/netdata/conf.d/apps_groups.conf' (errno 2, No such file or directory)
2021-11-30 14:55:53: apps.plugin INFO  : MAIN : Cannot read process groups configuration file '/etc/netdata/conf.d/apps_groups.conf'. Will try '/nix/store/7p35219vghilw3zp5v0zd2j1anyj5qyj-netdata-1.31.0/lib/netdata/conf.d/apps_groups.conf'
2021-11-30 14:55:53: apps.plugin INFO  : MAIN : Loaded config file '/nix/store/7p35219vghilw3zp5v0zd2j1anyj5qyj-netdata-1.31.0/lib/netdata/conf.d/apps_groups.conf'
2021-11-30 14:55:53: apps.plugin INFO  : MAIN : started on pid 60619
...
2021-11-30 14:55:53: go.d INFO: main[main] looking for 'cockroachdb.conf' in [/etc/netdata/conf.d/go.d /nix/store/7p35219vghilw3zp5v0zd2j1anyj5qyj-netdata-1.31.0/lib/netdata/conf.d/go.d]
2021-11-30 14:55:53: go.d INFO: main[main] found '/nix/store/7p35219vghilw3zp5v0zd2j1anyj5qyj-netdata-1.31.0/lib/netdata/conf.d/go.d/cockroachdb.conf
...
```

```diff
diff --git a/nixos/modules/services/monitoring/netdata.nix b/nixos/modules/services/monitoring/netdata.nix
index 3b3db7c9133..1f028cd4341 100644
--- a/nixos/modules/services/monitoring/netdata.nix
+++ b/nixos/modules/services/monitoring/netdata.nix
@@ -21,7 +21,6 @@ let

   configDirectory = pkgs.runCommand "netdata-config-d" { } ''
     mkdir -p $out
-    cp -r ${cfg.package}/lib/netdata/conf.d/* $out
     chmod -R +w $out
     ${concatStringsSep "\n" (mapAttrsToList (path: file: ''
         mkdir -p "$out/$(dirname ${path})"

```

**Comment by Izorkin** [comment]
Needed add `pgrep` command to `PATH`:
```
/nix/store/7p35219vghilw3zp5v0zd2j1anyj5qyj-netdata-1.31.0/libexec/netdata/plugins.d/system-info.sh: line 373: pgrep: command not found
2021-11-30 14:55:53: netdata INFO  : MAIN : NETDATA_CONTAINER_OS_NAME=none
```

**Comment by misuzu** [comment]
> Needed add `pgrep` command to `PATH`:
> 

Added `procps` to `PATH`.

**Review by Izorkin — APPROVED** [review]

File: nixos/modules/services/monitoring/netdata.nix [hunk 3]
```diff
@@ -150,20 +178,27 @@ in {
...
         ++ lib.optional cfg.python.enable (pkgs.python3.withPackages cfg.python.extraPackages)
         ++ lib.optional config.virtualisation.libvirtd.enable (config.virtualisation.libvirtd.package);
       environment = {
         PYTHONPATH = "${cfg.package}/libexec/netdata/python.d/python_modules";
       } // lib.optionalAttrs (!cfg.enableAnalyticsReporting) {
         DO_NOT_TRACK = "1";
       };
+      restartTriggers = [
+        config.environment.etc."netdata/netdata.conf".source
+        config.environment.etc."netdata/conf.d".source
```
Comment (wmertens) [inline nixos/modules/services/monitoring/netdata.nix:198]: won't these be store paths, so they will never change?

**Comment by wmertens** [comment]
looks ok to me, there's only one netdata so it doesn't hurt that it symlinks in /etc and it's nice for debugging/tweaking before updating the configuration. :+1:

File: nixos/modules/services/monitoring/netdata.nix (hunk 3 above)
Comment (misuzu) [reply nixos/modules/services/monitoring/netdata.nix:198]: `restartTriggers` accepts list of derivations.

Comment (wmertens) [reply nixos/modules/services/monitoring/netdata.nix:198]: I learned now that `restartTriggers` is a NixOS setting, not a systemd feature. Sorry for the noise!

**Comment by wmertens** [comment]
Let's just merge this? Seems safe?

**Review by wmertens — APPROVED** [review]

**Comment by wmertens** [comment]
Yup this looks great - any objections to me merging right now?

**Pull request merged**